### PR TITLE
[WIP] [SPARK-25603][SQL] Pushing Down Nested Field projections

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -611,6 +611,29 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Can't prune the columns on LeafNode
     case p @ Project(_, _: LeafNode) => p
 
+    // If the current project or the child references to nested fields, we can substitute them
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Can we please move this into a separate rule?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -611,6 +611,29 @@ object ColumnPruning extends Rule[LogicalPlan] {
...
+    // by alias attributes; then a project of the nested fields as aliases on the children
+    // of the child will be created.
+    //
+    // Note that if the child is also a [[Project]], it will be handled by the previous rule.
+    // If the child is a [[Filter]], it will be conflict with PushPredicatesThroughProject.
+    // When the child is a [[SerializeFromObject]], we can not push the projection through it.
+    case p @ Project(_, child) if !child.isInstanceOf[Project] && !child.isInstanceOf[Filter] &&
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I am generally not a big fan of blacklisting, we have seen various optimizer bugs caused by blacklisting in the past. Can we just create a white list of operators?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 1 above)
Comment (dbtsai) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I feel it should be part of `ColumnPruning`. 

Do you mean move this particular rule out or move the entire `ColumnPruning` into a new file? If the later, I chatted with @gatorsmile, and he felt moving the entire `ColumnPruning` out will lost the history which is not ideal.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (dbtsai) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I was debating the same. Will change to a white list of operators.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala [hunk 3]
```diff
@@ -85,10 +90,95 @@ class ParquetSchemaPruningSuite
...
+
+    val query = data.as[Contact]
+      .map(c => c.copy(id = 2))
+      .select("id", "name", "address", "friends")
+      .select(col("id"), firstAndLastName(col("name.first"), col("name.last")))
+
+    query.explain(true)
+    query.show(true)
+
+    val a =10
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaPruningSuite.scala]: Could you remove this line (which is failing the ScalaStyle test)? Then, we can see UT test result.

**Comment by maropu** [comment]
What's the status of this pr? just waiting for more reviews?

**Comment by dongjoon-hyun** [comment]
Thank you for attention, @maropu .

@dbtsai and I decided that I'm going to take over by making another PR based on this PR. @dbtsai will be the main author and I'll do some refactoring and proceed based on up-coming reviews. The deadline is before Spark Summit 2019.

**Comment by maropu** [comment]
Aha, I see. Thanks for the explanation, @dongjoon-hyun!

**Comment by dongjoon-hyun** [comment]
I close this WIP PR in favor of #23964 .

**Pull request closed without merging**